src/hotspot/share/oops/symbol.cpp

```diff
@@ -65,38 +65,10 @@ Symbol::Symbol(const u1* name, int length, int refcount) {
   memcpy(_body, name, length);
 }
 
-void* Symbol::operator new(size_t sz, int len) throw() {
-#if INCLUDE_CDS
- if (DumpSharedSpaces) {
-   MutexLocker ml(DumpRegion_lock, Mutex::_no_safepoint_check_flag);
-   // To get deterministic output from -Xshare:dump, we ensure that Symbols are allocated in
-   // increasing addresses. When the symbols are copied into the archive, we preserve their
-   // relative address order (sorted, see ArchiveBuilder::gather_klasses_and_symbols).
-   //
-   // We cannot use arena because arena chunks are allocated by the OS. As a result, for example,
-   // the archived symbol of "java/lang/Object" may sometimes be lower than "java/lang/String", and
-   // sometimes be higher. This would cause non-deterministic contents in the archive.
-   DEBUG_ONLY(static void* last = 0);
-   void* p = (void*)MetaspaceShared::symbol_space_alloc(size(len)*wordSize);
-   assert(p > last, "must increase monotonically");
-   DEBUG_ONLY(last = p);
-   return p;
- }
-#endif
-  int alloc_size = size(len)*wordSize;
-  address res = (address) AllocateHeap(alloc_size, mtSymbol);
-  return res;
-}
-
-void* Symbol::operator new(size_t sz, int len, Arena* arena) throw() {
-  int alloc_size = size(len)*wordSize;
-  address res = (address)arena->AmallocWords(alloc_size);
-  return res;
-}
-
-void Symbol::operator delete(void *p) {
-  assert(((Symbol*)p)->refcount() == 0, "should not call this");
-  FreeHeap(p);
+Symbol::Symbol(const Symbol& s1) {
```

Comment:
Can you add a comment that this copies the symbol when added to the ConcurrentHashTable.

Reply:
I've added a comment and have pushed a commit based on your and Ioi's comments.
For now, I've changed the parameter name in `do_add_if_needed()` from `heap` to `is_permanent` and adjusted the callsites accordingly.